subsys/mgmt/smp_udp.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2019, Prevas A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief UDP transport for the mcumgr SMP protocol.
+ */
+
+#include <zephyr.h>
+#include <init.h>
+#include <net/socket.h>
+#include <errno.h>
+
+#include <mgmt/smp_udp.h>
+#include <mgmt/buf.h>
+
+#include <mgmt/smp.h>
+
+static struct zephyr_smp_transport smp_udp_transport;
+static int smp_udp_receive_sock;
+static int smp_udp_send_sock;
+static u8_t receive_buf[CONFIG_MCUMGR_SMP_UDP_MTU];
+static struct k_thread smp_udp_thread;
+static K_THREAD_STACK_DEFINE(smp_udp_stack, CONFIG_MCUMGR_SMP_UDP_STACK_SIZE);
+static K_SEM_DEFINE(smp_udp_abort_sem, 0, 1);
+
+static int smp_udp_tx_pkt(struct zephyr_smp_transport *zst, struct net_buf *nb)
+{
+	ARG_UNUSED(zst);
+	int ret;
+	struct sockaddr_in *addr = net_buf_user_data(nb);
+
+	ret = zsock_sendto(smp_udp_send_sock, nb->data, nb->len, 0,
+			   (struct sockaddr *)addr, sizeof(*addr));
+	ret = ret < 0 ? errno : 0;
+
+	mcumgr_buf_free(nb);
+	return ret;
+}
+
+static uint16_t smp_udp_get_mtu(const struct net_buf *nb)
+{
+	return CONFIG_MCUMGR_SMP_UDP_MTU;
+}
+
+static int smp_udp_ud_copy(struct net_buf *dst, const struct net_buf *src)
+{
+	struct sockaddr_in *src_ud = net_buf_user_data(src);
+	struct sockaddr_in *dst_ud = net_buf_user_data(dst);
+
+	*dst_ud = *src_ud;
+
+	return 0;
+}
+
+static void smp_udp_receive_thread(void *p1, void *p2, void *p3)
+{
+	while (1) {
+		struct sockaddr addr;
+		socklen_t addr_len = sizeof(addr);
+
+		int len = zsock_recvfrom(smp_udp_receive_sock, receive_buf,
+					 CONFIG_MCUMGR_SMP_UDP_MTU, 0, &addr,
+					 &addr_len);
+
+		if (len > 0) {
+			struct sockaddr_in *ud;
+			struct net_buf *nb;
+
+			nb = mcumgr_buf_alloc();
+			net_buf_add_mem(nb, receive_buf, len);
+
+			ud = net_buf_user_data(nb);
+			*ud = *net_sin(&addr);
+
+			zephyr_smp_rx_req(&smp_udp_transport, nb);
+		} else if (len < 0) {
+			printk("smp: recvfrom error: %i", errno);
+		}
+
+		if (k_sem_take(&smp_udp_abort_sem, K_NO_WAIT) == 0) {
```

Comment:
Wouldn't volatile global lifetime variable suffice here instead of semaphore? What is benefit of using semaphore here?